### DISCUSSION TITLE
Feature Request: Triangulation+

### DISCUSSION THREAD

**Discussion by julianrinaldi** [mKeRix/room-assistant#401] [Ideas]
This is probably a very advanced request, but it would be amazing if there were a way to "calibrate" Room Assistant by having some sort of method to walk around a room and press a "set" button to register that you're in a certain room. Then Room Assistant could figure out where you are by measuring signal strength from multiple nodes.

My home layout is a bit difficult to calibrate everything well, because it has 5 levels, so I have many Raspberry Pi Zeros running at low BLE detection distances so they don't pick up Raspberry Pi Zeros from other levels. It works ok, but often gets my room wrong in some areas, even after a ton of tweaking.

I'm sure this would be a very difficult to implement, but it would make everything much more reliable.

**Comment by jaburges** [comment]
you could do this yourself with co-ordinates. I just wrote up a post about it :)
https://community.home-assistant.io/t/internal-position-tracking-with-co-ordinates-using-an-iphone-and-room-assistant/254399

**Reply by julianrinaldi** [reply]
Interesting, I set it up, but I'm not sure how useful it will be for locating people in 3D. I found this though:

https://www.npmjs.com/package/lm-trilateration3d

It looks like it might be more robust. Not exactly sure about how to use it but I might look into it.

**Reply by jaburges** [reply]
3D will be infinitely more complex (well not infinitely, an extra dimension, but you get what I mean).

My thought is that based on the floor you are on (some other sensor) you could triangulate on each floor, but with 5 levels, that's going to be tough. 

Depending on the distance and the bluetooth accuracy across the house (interference with 2.4GHz WiFi might make it painful) you may be able to accomplish it

**Reply by jaburges** [reply]
Also may try this over the weekend and see how whole house coverage is in 3D mapping :)

**Reply by jaburges** [reply]
seems to work pretty well - just need to space out the pis a little better. 
My understanding on the maths is the values of the x,y,z co-ordinates dont have to be perfect, but will skew the shape of the space if their positions aren't to scale.

For example if you have one device on the floor, and another on a table - lets say you mark that as a Y different of 2 (as i'm using imaginary feet as a measure) - however if you have one 2 floors above calculating that exactly will be hard. So i'm assuming a guess of 15 ft will be ok, but the way the formula works is there will be distortion of your spaces based on the inaccuracy of the scale.

Wondering whether using a floorplan app, then some hyper accurate altimeter briefly might help?

**Comment by jdiegmueller** [comment]
This is a pretty strong start!  I started down a similar path last week by
spinning up a Node-RED (had never used it; very neat tool) but wasn't smart
enough to implement any sort of trilateration; I was just trying to average
the RSSI of 2 sensors per room, and that didn't end up effective enough.

This looks like it certainly could pull it off, though.  I'll kick the
tires on this in the next week or so.  I do have CAD floorplans of the
house, so I should be able to get the coordinates dialed in accurately.

I'm curious where you're finding it best to locate the sensors?  I am no RF
expert, but my thought was that the best spot might be in the ceilings in
the middle of rooms such that a beacon always has LOS to a sensor since
they're "up".  I'll try a couple of different things once I get a chance to
kick the tires on this and let you know what I find!

**Reply by jaburges** [reply]
My expertise is bringing everything together so i know a little about a lot, sadly i dont know a lot about the specifics - so RF for example will be me bumbling around until I get it right lol.

Right now i have 2 on each floor, N & S ends on the top, E & W on the bottom floor. No clue if thats optimal, but will report back.
Glad to have someone else trying it as well.

If you're on discord hit me up with progress

**Reply by jdiegmueller** [reply]
@jaburges , how do I find you on Discord?  I've made quite a bit of progress this weekend.  I believe with some polish, this approach (lots of sensors + the triangluation library) seems like it's going to be a winner.

**Reply by jaburges** [reply]
JB#0050

**Comment by mKeRix** [comment]
Really cool to see you all experimenting with this! I was playing with the thought of doing something like this using neural networks, where you can train your setup similar to how you described above and it would then estimate the location. Kind of like how FIND3 does it - which is also why this was primarily discussed in #291 up until now.

I've already looked at this a little bit, but then got sidetracked with the iPhone BLE stuff. I still think this would be a really cool feature for the future though.
